src/main/java/com/melon/portfoliomanager/services/StockPricesService.java

```diff
@@ -0,0 +1,74 @@
+package com.melon.portfoliomanager.services;
+
+import com.melon.portfoliomanager.dtos.responses.StockPricesDTO;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.ResponseEntity;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Service
+public class StockPricesService {
+    private final double PRICE_CHANGE_PERCENT_THRESHOLD = 5;
+    private final Map<String, Double> companyStocks = new HashMap<>();
+    private final HttpService httpService;
+    private final NotificationService notificationService;
+
+    @Value("${stock.api.host}")
+    private String stockApiHost;
+
+    @Value("${stock.api.port}")
+    private String stockApiPort;
+
+    @Autowired
+    public StockPricesService(HttpService httpService, NotificationService notificationService) {
+        this.httpService = httpService;
+        this.notificationService = notificationService;
+    }
+
+    @Scheduled(fixedDelay = 15 * 60 * 1000)
+    private void notifyUserAboutExtremePriceChanges() {
```

Comment:
This method is doing a lot of different things. It calls an external stock price API, persists the response in memory, tracks for extreme price changes (which is business logic) and notifies users if it detects any. According to its name it only does the last part. Consider renaming it, or better yet - splitting up these responsibilities somehow.

Reply:
But creating an abstraction which will handle the internal state is something I was wondering if it will be justified enough in the moment in order to be created. What do you think ?

Reply:
I'm fine with the method staying as-is but not sure about the name. I'd say the main purpose of the scheduled code is to fetch stock price data regularly to keep it fresh, not notify users.

I think it would be better to have the scheduled method only fetch data and store the results. I agree with having some separate code do the actual updates. Then we can think about whose job it should be to track the changes and publish in Kafka - the class that manages the internal state is a good candidate. Maybe we can use PropertyChangeListener or an Observer pattern to add code that checks the difference and only gets called whenever the internal state changes (but this seems like overkill). But it's definitely not the scheduled method's job to do this. What if we want to add another way to update the internal state? We wouldn't be able to leverage the price change detection code and the users wouldn't be notified because everything is coupled with the scheduled code that gets the most recent prices.

I'm not saying this is the best approach, perhaps you have a better idea but I think it would be a nice exercise to try to improve this. The extra effort would certainly be justified, since the goal of the project is for you to get the most out of its development, not reach the end as quickly as possible. So I'm leaving the final decision to you - I'm fine with keeping it as-is too.

Side-note 1) We need to make sure we're not calling the notification service with empty lists every 15 minutes
Side-note 2) We must figure out which users should be notified

Reply:
I will try to make the changes in order to decouple internal state with the current implementation of fetching stock prices. I have intentionally coupled them so that nobody is able to mess up with the internal state - I think that only the scheduled method should have the rights to make changes to the internal state since this is what we need at the moment. May be I should also think about easier replacement of the stock prices API that we use because now we do not keep up very well with the open-closed principle and this replacement is something that is possible to happen if the project goes on.
Thank you for the feedback ! I will try to think of a better implementation.